File: samples/sample-app-storage/src/main/java/com/azure/android/storage/sample/TransferObserver.java
// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License.s

package com.azure.android.storage.sample;

import androidx.lifecycle.Observer;

import com.azure.android.storage.blob.transfer.TransferInfo;

/**
 * A convenient LiveData Observer for transfer state and progress changes.
 * The callback methods in this Observer will be invoked on the main thread.
 */
public interface TransferObserver extends Observer<TransferInfo> {
    /**
     * Called when transfer is accepted by system.
     *
     * @param transferId the transfer id
     */
    void onStart(String transferId);

    /**
     * Called when transfer made some progress by transferring by bytes.
     *
     * @param transferId the transfer id
     * @param totalBytes the total bytes to transfer
     * @param bytesTransferred the bytes transferred so far
     */
    void onProgress(String transferId, long totalBytes, long bytesTransferred);

    /**
     * Called when the transfer is paused by the system.
     *
     * @param transferId the transfer id
     */
    void onSystemPaused(String transferId);

    /**
     * Called when the paused transfer is resumed.
     *
     * @param transferId the transfer id
     */
    void onResume(String transferId);

    /**
     * Called when the transfer is completed.
     *
     * @param transferId the transfer id
     */
    void onComplete(String transferId);

    /**
     * Called when an error happens.
     *
     * @param transferId the transfer id
     * @param errorMessage the error message
     */
    void onError(String transferId, String errorMessage);

    @Override
    default void onChanged(TransferInfo transferInfo) {
        @TransferInfo.State int state = transferInfo.getState();
        switch (state) {
            case TransferInfo.State.STARTED:
                this.onStart(transferInfo.getId());
                break;
            case TransferInfo.State.RECEIVED_PROGRESS:
                TransferInfo.Progress progress = transferInfo.getProgress();
                this.onProgress(transferInfo.getId(), progress.getTotalBytes(), progress.getBytesTransferred());
                break;
            case TransferInfo.State.SYSTEM_PAUSED:
                this.onSystemPaused(transferInfo.getId());
                break;
            case TransferInfo.State.RESUMED:
                this.onResume(transferInfo.getId());
                break;
            case TransferInfo.State.COMPLETED:
                this.onComplete(transferInfo.getId());
                break;
            case TransferInfo.State.FAILED:
                this.onError(transferInfo.getId(), "Transfer failed.");
                break;
        }
    }
}

File: samples/sample-app-storage-kotlin/src/main/java/com/azure/android/storage/sample/kotlin/TransferObserver.kt
package com.azure.android.storage.sample.kotlin

import androidx.lifecycle.Observer
import com.azure.android.storage.blob.transfer.TransferInfo

/**
 * A convenient LiveData Observer for transfer state and progress changes.
 * The callback methods in this Observer will be invoked on the main thread.
 */
interface TransferObserver : Observer<TransferInfo?> {
    /**
     * Called when transfer is accepted by system.
     *
     * @param transferId the transfer id
     */
    fun onStart(transferId: Long)

    /**
     * Called when transfer made some progress by transferring by bytes.
     *
     * @param transferId the transfer id
     * @param totalBytes the total bytes to transfer
     * @param bytesTransferred the bytes transferred so far
     */
    fun onProgress(transferId: Long, totalBytes: Long, bytesTransferred: Long)

    /**
     * Called when the transfer is paused by the system.
     *
     * @param transferId the transfer id
     */
    fun onSystemPaused(transferId: Long)

    /**
     * Called when the paused transfer is resumed.
     *
     * @param transferId the transfer id
     */
    fun onResume(transferId: Long)

    /**
     * Called when the transfer is completed.
     *
     * @param transferId the transfer id
     */
    fun onComplete(transferId: Long)

    /**
     * Called when an error happens.
     *
     * @param transferId the transfer id
     * @param errorMessage the error message
     */
    fun onError(transferId: Long, errorMessage: String?)
    override fun onChanged(transferInfo: TransferInfo?) {
        when (transferInfo?.state) {
            TransferInfo.State.STARTED -> onStart(transferInfo.id)
            TransferInfo.State.RECEIVED_PROGRESS -> {
                val progress = transferInfo.progress
                onProgress(transferInfo.id, progress.totalBytes, progress.bytesTransferred)
            }
            TransferInfo.State.SYSTEM_PAUSED -> onSystemPaused(transferInfo.id)
            TransferInfo.State.RESUMED -> onResume(transferInfo.id)
            TransferInfo.State.COMPLETED -> onComplete(transferInfo.id)
            TransferInfo.State.FAILED -> onError(transferInfo.id, "Transfer failed.")
        }
    }
}
